### PR TITLE
implement Fluidsynth music backend

### PR DESCRIPTION
**Pull request by rfomin**

### CODE REVIEW

**Comment by rfomin** [comment]
I have a few questions:
1. Currently fluidsynth backend is optional and works only with MSYS2/Linux builds. We can make it mandatory and add the download of fluidsynth libraries from the official release for MSVC/cross-compilation. Or at least search for it on MSVC (currently FindFluidSynth.cmake uses pkgconfig).
2. It seems `-nosfx` parameter is broken because we don't initialize SDL audio in music backends (except opl). Should I add SDL audio initialization to `i_sdlmusic` and `i_flmusic`?

**Comment by fabiangreffrath** [comment]
Sorry for the late reply! I have seen that you asked two questions which were since then removed, but I'd like to give my opinion anyway:

> Currently fluidsynth backend is optional and works only with MSYS2/Linux builds. We can make it mandatory and add the download of fluidsynth libraries from the official release for MSVC/cross-compilation. Or at least search for it on MSVC (currently FindFluidSynth.cmake uses pkgconfig).

Yes, let's please keep fluidsynth as a strictly optional dependency. Music playback works without it and I don't want to force this heavy-weight dependency on everyone. Let's keep this port as light as possible.

> It seems -nosfx parameter is broken because we don't initialize SDL audio in music backends (except opl). Should I add SDL audio initialization to i_sdlmusic and i_flmusic?

Yes, let's please fix the `-nosfx` parameter. It is strictly supposed to work independent of the music backend.

File: Source/i_sound.c [hunk 1]
```diff
@@ -678,13 +684,7 @@ void I_ShutdownSound(void)
...
 void I_InitSound(void)
 {   
-   // [FG] initialize music backend function pointers
-   if (music_backend == music_backend_opl)
-      I_OPL_InitMusicBackend();
-   else
-      I_SDL_InitMusicBackend();
-
-   if(!nosfxparm)
+   if(!nosfxparm || !nomusicparm)
```
Comment (rfomin) [inline Source/i_sound.c:687]: > Yes, let's please fix the `-nosfx` parameter. It is strictly supposed to work independent of the music backend.

I fixed it with this change. Not sure if it's worth adding SDL initialization to every music module like in Choco, I think it's not necessary.

File: Source/i_sound.h [hunk 2]
```diff
@@ -90,39 +92,59 @@ int I_SoundID(int handle);
...
+    void (*I_UnRegisterSong)(void *handle);
+} music_module_t;
+
 typedef enum
 {
-  music_backend_sdl,
-  music_backend_opl,
-  num_music_backends,
-} music_backend_t;
+#if defined(_WIN32)
```
Comment (fabiangreffrath) [inline Source/i_sound.h:110]: No SDL backend on WIndows? How do we playback MP3 music then?

File: Source/i_sound.c [hunk 3]
```diff
@@ -740,17 +740,94 @@ void I_InitSound(void)
       // haleyjd 04/11/03: don't use music if sfx aren't init'd
       // (may be dependent, docs are unclear)
       if(!nomusicparm)
-         I_InitMusic();
+      {
```
Comment (fabiangreffrath) [inline Source/i_sound.c:743]: Don't we have to initialize SDL_Mixer even in the `nomusic` case for the SFX to work?

File: Source/i_winmusic.c [hunk 4]
```diff
@@ -437,7 +440,34 @@ void I_WIN_PlaySong(boolean looping)
...
+    MMRESULT mmr;
+
+    mmr = midiStreamRestart(hMidiStream);
+    if (mmr != MMSYSERR_NOERROR)
+    {
+        MidiErrorMessageBox(mmr);
+    }
+}
+
+static boolean IsMid(byte *mem, int len)
```
Comment (fabiangreffrath) [inline Source/i_winmusic.c:465]: I think I have seen this function the fourth time now. Maybe we should move it into `i_sound.h`?

File: Source/m_misc.c [hunk 5]
```diff
@@ -1961,19 +1965,36 @@ default_t defaults[] = {
...
-    (config_t *) &fluidsynth_sf_path, NULL,
-    {SFINIT(.s, "")}, {0}, string, ss_none, wad_no,
+    "soundfont_path",
+    (config_t *) &soundfont_path, NULL,
+    {SFINIT(.s, "soundfonts"DIR_SEPARATOR_S"TimGM6mb.sf2")},
+    {0}, string, ss_none, wad_no,
     "FluidSynth soundfont path"
   },
 
+  {
```
Comment (fabiangreffrath) [inline Source/m_misc.c:1983]: Do we really need to set chorus and reverb? Should we just keep them disabled?

File: Source/i_sound.h (hunk 2 above)
Comment (rfomin) [reply Source/i_sound.h:110]: The SDL music backend is now always initialized, `I_RegisterSong()` calls the SDL module for non-MIDI music. See `I_RegisterSong` function in `i_sound.c`.

File: Source/i_sound.c (hunk 3 above)
Comment (rfomin) [reply Source/i_sound.c:743]: We initialize it earlier in this function. I have tested that `-nosfx` and `-nomusic` now work correctly, at least on Windows.

File: Source/m_misc.c (hunk 5 above)
Comment (rfomin) [reply Source/m_misc.c:1983]: Personally, I don't like reverb and chorus, but maybe someone needs it. PrBoom+ and GZDoom can enable it.

File: Source/i_sound.h (hunk 2 above)
Comment (fabiangreffrath) [reply Source/i_sound.h:110]: 👍

File: Source/i_sound.c (hunk 3 above)
Comment (fabiangreffrath) [reply Source/i_sound.c:743]: 👍

Comment (fabiangreffrath) [reply Source/i_sound.c:743]: 👍

File: Source/m_misc.c (hunk 5 above)
Comment (fabiangreffrath) [reply Source/m_misc.c:1983]: Okay, I don't like them either, that's why I am asking. 😄 But I agree it won't hurt to leave them as a choice.

File: Source/i_winmusic.c (hunk 4 above)
Comment (rfomin) [reply Source/i_winmusic.c]: Done in https://github.com/fabiangreffrath/woof/pull/403/commits/bccf5e796d7786eb60ae8c26231eef697a30fdc7

**Review by fabiangreffrath — APPROVED** [review]

File: Source/i_sound.h (hunk 2 above)
Comment (fabiangreffrath) [reply Source/i_sound.h:110]: Wait a minute! Does this mean we can play back MIDI music with OPL emulation and the next song is in MP3 format and we can play it back with SDL? How cool is that?!

Comment (rfomin) [reply Source/i_sound.h:110]: Yes, we can. Maybe we should add this to Crispy :smile:

**Comment by fabiangreffrath** [comment]
Is this ready for merging?

**Comment by rfomin** [comment]
> Is this ready for merging?

Yes, it's ready.

**Comment by fabiangreffrath** [comment]
Great job, thank you!

File: CMakeLists.txt [hunk 6]
```diff
@@ -141,9 +138,14 @@ if(NOT SDL2_NET_FOUND)
...
         endif()
         set(SDL2_NET_DIR "${CMAKE_CURRENT_BINARY_DIR}/SDL2_net-2.0.1")
-        find_package(SDL2_net)
     endif()
 endif()
+find_package(SDL2_net)
+
+find_package(FluidSynth)
+if(FluidSynth_FOUND)
+    install(DIRECTORY soundfonts/ DESTINATION soundfonts)
```
Comment (rfomin) [inline CMakeLists.txt:147]: @fabiangreffrath This line packages soundfont on WIN32, I don't know what it does on *NIX.

**Pull request merged**